Additional constraints (#975)
CadQuery/cadquery: doc/assy.rst

> For the cost function of Plane, please see the Point and Axis sections. The ``param`` argument is
> copied into both constraints and should be left as the default value of ``None`` for a "mate" style

Reviewer: ```suggestion
For the cost function of Plane, please see the Point and Axis sections. The ``param`` argument is applied to Axis and should be left as the default value for a "mate" style
```

The cost function is:

.. math::

   \left\lVert \vec{ c } - \vec{param} \right\rVert ^2

Where:

- :math:`\vec{ c }` is the center of the argument,
- :math:`param` is the parameter of the constraint - tuple specifying the target position,
- :math:`\operatorname{dist}( \vec{ a }, b)` is the distance between point :math:`\vec{ a }` and
  line :math:`b`.

.. cadquery::

    import cadquery as cq

    b1 = cq.Workplane().box(1,1,1)
    b2 = cq.Workplane().sphere(0.15)

    assy = (
        cq.Assembly()
        .add(b1,name='b1')
        .add(b2, loc=cq.Location(cq.Vector(0,0,4)), name='b2', color=cq.Color('red'))
    )

    # fix the position of b1
    assy.constrain('b1','Fixed')
    # b2 on one of the edges of b1
    assy.constrain('b2','b1@edges@>>Z and >>Y','PointOnLine')
    # b2 on another of the edges of b1
    assy.constrain('b2','b1@edges@>>Z and >>X','PointOnLine')
    # effectively b2 will be constrained to be on the intersection of the two edges

    assy.solve()
    show_object(assy)

FixedRotation
=============

> FixedRotation fixes the rotation of the given argument to be equal to the given point specified via the parameter of the constraint. This constraint locks all rotational degrees of freedom of the argument.

Reviewer 2: ```suggestion
FixedRotation fixes the rotation of the given argument to be equal to the value specified via the parameter of the constraint. The argument is rotated about it's origin firstly by the Z angle, then Y and finally X.

This constraint locks all rotational degrees of freedom of the argument.
```
At least I think that's correct, please do check closely @adam-urbanczyk.

FixedRotation fixes the rotation of the given argument to be equal to the value specified via the parameter of the constraint. The argument is rotated about it's origin firstly by the Z angle, then Y and finally X.

This constraint locks all rotational degrees of freedom of the argument.
The cost function is:

.. math::

   \left\lVert \vec{ R } - \vec{param} \right\rVert ^2

Where:

- :math:`\vec{ R }` vector of the rotation angles of the rotation applied to the argument,
- :math:`param` is the parameter of the constraint - tuple specifying the target rotation.

.. cadquery::

    import cadquery as cq

    b1 = cq.Workplane().box(1,1,1)
    b2 = cq.Workplane().rect(0.1, 0.1).extrude(1,taper=-15)

    assy = (
        cq.Assembly()
        .add(b1,name='b1')
        .add(b2, loc=cq.Location(cq.Vector(0,0,4)), name='b2', color=cq.Color('red'))
    )

    # fix the position of b1
    assy.constrain('b1','Fixed')
    # fix b2 bottom face position (but not rotation)
    assy.constrain('b2@faces@<Z','FixedPoint',(0,0,0.5))
    # fix b2 rotational degrees of freedom too
    assy.constrain('b2','FixedRotation',(45,0,45))

    assy.solve()
    show_object(assy)

FixedAxis
=========

> FixedAxis fixes the orientation of the given argument's normal or tangent to be equal to the orientation  of the vector specified via the parameter of the constraint. This constraint locks two rotational degrees of freedom of the argument.

Reviewer 3: ```suggestion
FixedAxis fixes the orientation of the given argument's normal or tangent to be equal to the orientation of the vector specified via the parameter of the constraint. This constraint locks two rotational degrees of freedom of the argument.
```
